base/array.jl

```diff
@@ -824,22 +824,23 @@ function collect_to_with_first!(dest, v1, itr, st)
 end
 
 function setindex_widen_up_to(dest::AbstractArray{T}, el, i) where T
-    @inline
     new = similar(dest, promote_typejoin(T, typeof(el)))
     f = first(LinearIndices(dest))
     copyto!(new, first(LinearIndices(new)), dest, f, i-f)
-    @inbounds new[i] = el
+    new[i] = el
     return new
 end
 
 function collect_to!(dest::AbstractArray{T}, itr, offs, st) where T
     # collect to dest array, checking the type of each result. if a result does not
     # match, widen the result type and re-dispatch.
     i = offs
+    len = length(dest)
     while true
         y = iterate(itr, st)
         y === nothing && break
         el, st = y
+        @assert len == length(dest)
```

Comment:
Is this really an assertion or should we give an error like "destination changed size during iteration; this is not supported"?

Reply:
If the code relies on this property, shouldn't this be more like

```
len == length(dest) || throw(...)
```

, since `@assert` is documented to not be used for critical checks of correctness?